transport/transporters/s3/transporter/transporter.go

```diff
@@ -0,0 +1,351 @@
+
+	if *endpoint != "" {
+		sess = session.Must(session.NewSession(&aws.Config{
+			Region:      		aws.String(*awsRegion),
+			Credentials: 		creds,
+			S3ForcePathStyle: 	aws.Bool(true),
+			Endpoint:    		aws.String(*endpoint),
+			MaxRetries:			aws.Int(0), // We disable the client retry policy because of our own retry logic
+		}))
+	} else {
+		sess = session.Must(session.NewSession(&aws.Config{
+			Region:      aws.String(*awsRegion),
+			Credentials: creds,
+			MaxRetries:	 aws.Int(0),
+		}))
+	}
+
+	client := s3.New(sess)
+
+	return NewTransporterWithInterface(
+		shutdownHandler,
+		inputChan,
+		txnsWritten,
+		statsChan,
+		log,
+		id,
+		bucketName,
+		keySpace,
+		client,
+		retryPolicy)
+}
+
+// shutdown idempotently closes the output channel
+func (t *S3Transporter) shutdown() {
+	t.log.Info("shutting down transporter")
+	t.shutdownHandler.CancelFunc() // initiate shutdown on other modules as well
+
+	if r := recover(); r != nil {
+		t.log.Warnf("Recovered in S3Transporter %s", r)
+	}
+
+	defer func() {
+		// recover if channel is already closed
+		recover()
+	}()
+
+	t.log.Debug("closing progress channel")
+	close(t.txnsWritten)
+}
+
+
+// transportWithRetry does a PUT on a full batch as a single key/file to S3
+func (t *S3Transporter) transportWithRetry(ctx context.Context, messagesSlice []*marshaller.MarshalledMessage) (error, bool) {
+	var cancelled bool
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+
+	select {
+	case <-ctx.Done():
+		t.log.Debug("received terminateCtx cancellation")
+		cancelled = true
+	default:
+	}
+
+	// add all messages into a gzipped buffer
+	for _, msg := range messagesSlice {
+		if _, err := gz.Write(msg.Json); err != nil {
+			return err, cancelled
+		}
+		gz.Write([]byte("\n"))
+	}
+
+	if err := gz.Flush(); err != nil {
+		return err, cancelled
+	}
+	if err := gz.Close(); err != nil {
+		return err, cancelled
+	}
+
+	gz.Reset(ioutil.Discard)
+
+	firstWalStart := messagesSlice[0].WalStart
+
+	// Free up messagesSlice now that we're done with it
+	messagesSlice = nil
+
+	// Clear out every allocation in the conversion
+	byteArray := buf.Bytes()
+	buf.Reset()
+
+	byteReader := bytes.NewReader(byteArray)
+	byteArray = nil
+
+	// Partition the S3 keys into days
+	year, month, day, full := TimeSource.DateString()
+	baseFilename := fmt.Sprintf("%s_%d", full, firstWalStart)
+
+	fullKey := key_join(true, t.keySpace, year, month, day, baseFilename)
+
+	// An operation that may fail.
+	operation := func() error {
+		// Do the upload and let S3 handle retries
+		_, err := t.client.PutObjectWithContext(ctx, &s3.PutObjectInput{
+			Bucket:          aws.String(t.bucketName),
+			Key:             aws.String(fullKey),
+			Body:            byteReader,
+			ContentEncoding: aws.String("gzip"),
+		})
+
+		// If any errors occurred during sending the entire batch
+		if err != nil {
+			t.log.WithError(err).Errorf("wal_start %d failed to be uploaded to S3", firstWalStart)
```

Comment:
Does it make sense to use `fullKey` instead of or in addition to `firstWalStart`?

Reply:
Sure, it provides more info.